Indicators/Vortex.cs

```diff
@@ -12,121 +13,133 @@
+
+        /// <summary>
+        /// Gets the Negative Vortex Indicator (-VI).
+        /// </summary>
+        public IndicatorBase<IndicatorDataPoint> MinusVI { get; private set; }
+
+        /// <summary>
+        /// Indicates whether the indicator is ready to produce meaningful results after its initial warm-up period.
+        /// </summary>
+        public override bool IsReady => Samples >= _period;
+
+        /// <summary>
+        /// Gets the required period, in data points, for the indicator to be considered ready and fully initialized.
+        /// </summary>
+        public int WarmUpPeriod => _period;
+
+        /// <summary>
+        /// Initializes a new instance of the Vortex Indicator with a specified period.
+        /// </summary>
+        public  Vortex(int period)
+            : this($"VTX({period})", period)
         {
-            return 0m; // or possibly Decimal.MinValue to indicate insufficient data
         }
 
-        decimal sumTR = 0m;
-        decimal sumVMPlus = 0m;
-        decimal sumVMMinus = 0m;
-
-        // Starting from 1 since we want to compare current bar with the previous bar
-        for (int i = 1; i < _window.Count; i++)
+        public Vortex(string name,int period)
+            : base(name)
         {
-            var current = _window[i - 1]; // current bar
-            var previous = _window[i]; // previous bar
+            _period = period;
+            _atr = new AverageTrueRange("VTX_ATR_" + period, period, MovingAverageType.Simple);
 
-            decimal tr = Math.Max(current.High - current.Low, Math.Max(Math.Abs(current.High - previous.Close), Math.Abs(current.Low - previous.Close)));
-            decimal vmPlus = Math.Abs(current.High - previous.Low);
-            decimal vmMinus = Math.Abs(current.Low - previous.High);
+            _plusVM = new List<decimal>();
+            _minusVM = new List<decimal>();
 
-            sumTR += tr;
-            sumVMPlus += vmPlus;
-            sumVMMinus += vmMinus;
-        }
-
-        decimal positiveVortex = sumVMPlus / sumTR;
-        decimal negativeVortex = sumVMMinus / sumTR;
-
-        PositiveVortexIndicator.Update(input.Time, positiveVortex);
-        NegativeVortexIndicator.Update(input.Time, negativeVortex);
-
-        // Depending on your test setup, you might want to return different values:
-        return positiveVortex; // If the test expects Positive Vortex
-        // return negativeVortex; // If the test expects Negative Vortex
-    }
-
-    /// <summary>
-    /// Gets the positive vortex indicator.
-    /// </summary>
-    public IndicatorBase<IndicatorDataPoint> PositiveVortexIndicator { get; } = new Identity("PositiveVortex");
-
-    /// <summary>
-    /// Gets the negative vortex indicator.
-    /// </summary>
-    public IndicatorBase<IndicatorDataPoint> NegativeVortexIndicator { get; } = new Identity("NegativeVortex");
+            PlusVI = new FunctionalIndicator<IndicatorDataPoint>("PlusVI",
```

Comment:
Take a look at our `IndicatorExtensions`. These could be composite indicators once you have `_atr`, `_atrSum` and the plus/minus sum indicators, by composing them like `PlusVortex = _plusSum.Over(_atrSum);`.

This will simplify your math and reduce your code by re-using what's already in Lean 👌